src/Data/Singletons/Prelude/Show.hs

```diff
@@ -0,0 +1,229 @@
+-- concatenation of results using function composition.
+type SymbolS = Symbol -> Symbol
+
+-- | GHC currently has no notion of type-level 'Char's, so we fake them with
+-- single-character 'Symbol's.
+type SChar = Symbol
+
+$(singletonsOnly [d|
+  class Show a where
+    showsPrec :: Nat -> a -> SymbolS
+    -- Intentionally renamed to avoid clashing with the Show class
+    show'     :: a -> Symbol
+    showList  :: [a] -> SymbolS
+
+    showsPrec _ x s = show' x <> s
+    show' x         = shows x ""
+    showList ls   s = showListWith shows ls s
+
+  shows :: Show a => a -> SymbolS
+  shows s = showsPrec 0 s
+
+  showListWith :: (a -> SymbolS) -> [a] -> SymbolS
+  showListWith _     []     s = "[]" <> s
+  showListWith showx (x:xs) s = "["  <> showx x (showl xs)
+    where
+      showl []     = "]" <> s
+      showl (y:ys) = "," <> showx y (showl ys)
+
+  showChar :: SChar -> SymbolS
+  showChar = (<>)
+
+  showString :: Symbol -> SymbolS
+  showString = (<>)
+
+  showParen :: Bool -> SymbolS -> SymbolS
+  showParen b p = if b then showChar "(" . p . showChar ")" else p
+
+  showSpace :: SymbolS
+  showSpace = \xs -> " " <> xs
+
+  showCommaSpace :: SymbolS
+  showCommaSpace = showString ", "
+
+  appPrec, appPrec1 :: Nat
+  appPrec  = 10
+  appPrec1 = 11
+
+  instance Show a => Show [a] where
+    showsPrec _ = showList
+
+  -- -| This is not an ideal Show instance for Symbol, since the Show instance
+  -- for String escapes special characters. Unfortunately, GHC lacks the ability
+  -- to case on individual characters in a Symbol (at least, not without GHC
+  -- plugins), so this is the best we can do for the time being.
+  instance Show Symbol where
+    showsPrec _ = showString
+
+  show_tuple :: [SymbolS] -> SymbolS
+  show_tuple ss = showChar "("
+                . foldr1 (\s r -> s . showChar "," . r) ss
+                . showChar ")"
+
+  instance (Show a, Show b) => Show (a,b)  where
+    showsPrec _ (a,b) s = show_tuple [shows a, shows b] s
+
+  instance (Show a, Show b, Show c) => Show (a, b, c) where
+    showsPrec _ (a,b,c) s = show_tuple [shows a, shows b, shows c] s
+
+  instance (Show a, Show b, Show c, Show d) => Show (a, b, c, d) where
+    showsPrec _ (a,b,c,d) s = show_tuple [shows a, shows b, shows c, shows d] s
+
+  instance (Show a, Show b, Show c, Show d, Show e) => Show (a, b, c, d, e) where
+    showsPrec _ (a,b,c,d,e) s = show_tuple [shows a, shows b, shows c, shows d, shows e] s
+
+  instance (Show a, Show b, Show c, Show d, Show e, Show f) => Show (a,b,c,d,e,f) where
+    showsPrec _ (a,b,c,d,e,f) s = show_tuple [shows a, shows b, shows c, shows d, shows e, shows f] s
+
+  instance (Show a, Show b, Show c, Show d, Show e, Show f, Show g)
+          => Show (a,b,c,d,e,f,g) where
+    showsPrec _ (a,b,c,d,e,f,g) s
+          = show_tuple [shows a, shows b, shows c, shows d, shows e, shows f, shows g] s
+  |])
+
+-- At the moment, I can't think of a way to implement a PShow Nat instance,
```

Comment:
I was afraid you'd ask that :)

I suppose I could puzzle my way to an implementation, if you'd want that. I opted not to in this iteration since the [value-level `Integer -> String`](https://github.com/ghc/ghc/blob/ba597c1dd1daf9643b72dc7aeace8d6b3fce84eb/libraries/base/GHC/Show.hs#L471-L548) function that GHC implements is too complicated to promote, and I'm not clever enough to figure out how to make a simpler version off the top of my head.